### ISSUE TITLE
Weighted highlights don't display newly visited pages until second new tab

### ISSUE THREAD

**Issue by dmose** [mozilla/activity-stream#1493]
steps to reproduce:  on latest master
1. `npm run once`
2. open a new tab
3. click on the yahoo icon
4. command-click on 5 or so stories so that they open in separate tabs
5. open a new tab
6. notice that there are (usually!) still only the three initial highlights, instead of highlights including the newly created history. (bug)
7. open another new tab
8. notice that this one renders something that includes the new history (correct)

**Comment by rlr** [comment]
I was able to reproduce this locally too. If you add step `7. open another new tab`, then they show up. Something with caching?

**Comment by dmose** [comment]
Interesting... could be that.  "git bisect" is returning weird results, which could be consistent with that, or with it being a race of some sort.  I'll try the step 7 thing.

**Comment by dmose** [comment]
I just saw a case where it worked correctly on step 6.  Which makes me think this is a race.

**Comment by dmose** [comment]
This has been around a while, it turns out.  At least since 64330436dd020285d34f457b7e59c107a7e34895 which is when the weightedHighlights experiment was forced on, quite possibly further back than that.  CC @piatra @sarracini

**Comment by dmose** [comment]
It appears (I've just tested a little), that 9d610d661475ad6cc978683181d1903fde5162ca does have this problem while 04055f4 does not.   So combined with @rlr's step 7 and some bisecting, we ought to be able to find where this was introduced.

**Comment by dmose** [comment]
There is a definitely some sort of race here, and it may indeed be a side effect of the caching architecture.  After opening 5 or 6 articles on yahoo as described above, I opened three new tabs, each about a second apart.  Here is what they look like:

![screen shot 2016-10-10 at 20 59 06](https://cloud.githubusercontent.com/assets/425657/19258426/a3743186-8f2c-11e6-869a-49f6105ceee6.png)

![screen shot 2016-10-10 at 20 59 31](https://cloud.githubusercontent.com/assets/425657/19258430/ad1e7124-8f2c-11e6-8573-628b4cd65142.png)

![screen shot 2016-10-10 at 20 59 42](https://cloud.githubusercontent.com/assets/425657/19258438/b6d26d6a-8f2c-11e6-9f2a-192e238d17bb.png)

This is on latest master, and nothing was done between those three screenshots, other than waiting.

**Comment by piatra** [comment]
I can have a look at this.

**Comment by piatra** [comment]
This works as designed/implemented.
I was able to reproduce this, it is not a race condition. It has to do with the way the caching mechanism works. Explanation (based on the STR above): when you open a new tab on step 5 you trigger the store refresh **that will _only_ benefit future new tabs**.
Key line in the code for this is https://github.com/mozilla/activity-stream/blob/575787869550a639f6571f0115eca5e80c79f675/addon/ActivityStreams.js#L691 
I don't think it's a deal breaker, usual browsing behavior (with some browsing history) will not suffer from this because they do not get such recent results.
cc @k88hudson

**Comment by dmose** [comment]
Thanks for digging in, @piatra!

Ok, so knowing that it doesn't work once history is primed is excellent news.  There is still the question of the first time user case, which in this case really means "a first time user of firefox who is also a first time user of activity stream", presumably mostly first-time users of Firefox once activity stream lands (though if there are people who download firefox and immediately install AS now, this would effect them too).

**Comment by Mardak** [comment]
Reopening as a "newly visited pages" issue

**Comment by k88hudson** [comment]
Closing in favour of resolving #1593 instead

**Comment by dmose** [comment]
I'm confused.  As nearly as I can tell, this is not related to #1593.  Reopening.  Let me know if I've misunderstood.

**Comment by dmose** [comment]
At the work week, @jaredkerim suggested that rather than pulling the metadata through the cache to fix this, which might be fragile, that we should consider doing what the non-weightedHeightlights code does: it simply won't show stuff without images.

I imagine the right way to do this in the new code would be to simply adjust the scoring of the stuff without images, but I haven't looked...

**Comment by k88hudson** [comment]
This should be fixed by #1669

**Comment by sarracini** [comment]
Fixed with #1677
